### PR TITLE
Fix and update Elixir generator

### PR DESCRIPTION
**Pull request by njouanin**
@mrmstn

Update Elixir generator with Tesla version and use Jason instead of Poison.
Also fix a bug when var name contains an '@'.

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [ ] In case you are adding a new generator, run the following additional script : 
  ```
  ./bin/utils/ensure-up-to-date
  ``` 
  Commit all changed files.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (6.3.0) (minor release - breaking changes with fallbacks), `7.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

**Comment by mrmstn** [comment]
Hi @njouanin 

Thank you very much for the PR. A switch to Jason would indeed be awesome.

Could you please also commit the created elixir samples from the checklist?

I'll take a more profound look into the PR on the weekend, many thanks for your patience

**Comment by barttenbrinke** [comment]
Came here to see if I could help out with the switch from Poison to Json as currently we are just using sed to replace Poison with Jason after a generate. 👍

File: modules/openapi-generator/src/main/resources/elixir/mix.exs.mustache [hunk 1]
```diff
@@ -4,7 +4,7 @@ defmodule {{moduleName}}.Mixfile do
   def project do
     [
       app: {{#atom}}{{#underscored}}{{packageName}}{{/underscored}}{{/atom}},
-      version: "{{appVersion}}",
+      version: "{{packageVersion}}",
```
Comment (mrmstn) [inline modules/openapi-generator/src/main/resources/elixir/mix.exs.mustache:7]: Samples will fail, due to `packageVersion` being undefined.

```
** (Mix) Expected :version to be a valid Version, got: "" (see the Version module for more information)
```

File: samples/client/petstore/elixir/mix.exs [hunk 2]
```diff
@@ -4,7 +4,7 @@ defmodule OpenapiPetstore.Mixfile do
   def project do
     [
       app: :openapi_petstore,
-      version: "1.0.0",
+      version: "",
```
Comment (mrmstn) [inline samples/client/petstore/elixir/mix.exs:7]: Compile error
```
** (Mix) Expected :version to be a valid Version, got: "" (see the Version module for more information)

```

File: modules/openapi-generator/src/main/resources/elixir/model.mustache [hunk 3]
```diff
@@ -4,34 +4,15 @@
   {{&description}}
   """
 
-  @derive [Poison.Encoder]
+  @derive Jason.Encoder
   defstruct [
-    {{#vars}}{{#atom}}{{&baseName}}{{/atom}}{{^-last}},
+    {{#vars}}{{#atom}}{{name}}{{/atom}}{{^-last}},
```
Comment (mrmstn) [inline modules/openapi-generator/src/main/resources/elixir/model.mustache:9]: This changes the name and writing of the payload, which results in breaking the spec due to diverging of the property name.

see generated examples.

File: modules/openapi-generator/src/main/resources/elixir/model.mustache [hunk 4]
```diff
@@ -4,34 +4,15 @@
...
     {{/-last}}{{/vars}}
   ]
 
   @type t :: %__MODULE__{
-    {{#vars}}{{#atom}}{{&baseName}}{{/atom}} => {{{datatype}}}{{#isNullable}} | nil{{/isNullable}}{{^isNullable}}{{^required}} | nil{{/required}}{{/isNullable}}{{^-last}},
+    {{#vars}}{{#atom}}{{name}}{{/atom}} => {{{datatype}}}{{#isNullable}} | nil{{/isNullable}}{{^isNullable}}{{^required}} | nil{{/required}}{{/isNullable}}{{^-last}},
```
Comment (mrmstn) [inline modules/openapi-generator/src/main/resources/elixir/model.mustache:14]: Same here, changing the properiename (or cases) breaks the openapi spec

File: samples/client/petstore/elixir/lib/openapi_petstore/model/_special_model_name_.ex [hunk 5]
```diff
@@ -6,19 +6,13 @@ defmodule OpenapiPetstore.Model.SpecialModelName do
   
   """
 
-  @derive [Poison.Encoder]
+  @derive Jason.Encoder
   defstruct [
-    :"$special[property.name]"
+    :dollar_special_left_square_bracket_property_period_name_right_square_bracket
```
Comment (mrmstn) [inline samples/client/petstore/elixir/lib/openapi_petstore/model/_special_model_name_.ex:11]: `$special[property.name]` is the expected payload name

**Review by mrmstn — CHANGES_REQUESTED** [review]
Hi @njouanin,

Thanks for your work on this PR. I've checked the client samples and noticed a few areas that need changes, especially around ensuring the payload variables aren't modified by the client.

Could you take a look at the comments I've made, address the issues, and push your updates?

Thanks again for your time and effort. Looking forward to getting this PR across the finish line 🚀

**Comment by mrmstn** [comment]
Hi @barttenbrinke,

Appreciate your review and for bringing this PR back into focus. I've made some comments on specific lines that require adjustments. 

If you're willing and able to contribute further, it'd be great if you could fork @njouanin's branch and make the necessary changes. 

@njouanin, would you be open to assisting @barttenbrinke if they decide to step in and work on this? 

Thanks again for everyone's collaboration and contributions ❤️. Looking forward to seeing this PR progress.

**Comment by barttenbrinke** [comment]
Ugh, Jason.Decoder no longer supports decoding into Structs by default (as it can be unsafe) - however as you are generating libs here, is not a concern - https://github.com/michalmuskala/jason#differences-to-poison Let me think about how to fix this.

**Comment by wing328** [comment]
From what I can tell, this PR includes 3 changes:

1. Update Elixir generator with Tesla version
2. use Jason instead of Poison.
3. Also fix a bug when var name contains an '@' 

What about filing 3 separate PRs instead for easier review and get these merged more quickly?

**Comment by mrmstn** [comment]
Hi @njouanin

Since this PR hasn't progressed over the past year and has significantly diverged from the upstream, I'll proceed with closing it for now.

Please feel free to reopen or resubmit whenever you're ready to revisit it. We'd be happy to review any future updates.

Thanks for your contributions!

**Comment by wing328** [comment]
@njouanin I would suggest breaking down this PR into smaller PRs if possible:

1. Update Elixir generator with Tesla version
2. use Jason instead of Poison.
3. Also fix a bug when var name contains an '@'.

**Pull request closed without merging**